src/mediarithmics/plugins/computed-field/ComputedFieldBasePlugin.ts

```diff
@@ -0,0 +1,104 @@
+import { BasePlugin } from '../common/BasePlugin';
+import * as ion from 'ion-js';
+import express from 'express';
+
+export type OnUpdatePluginStatus = 'ok' | 'error';
+
+export interface PluginResponse {
+  status: OnUpdatePluginStatus;
+}
+export interface OnUpdatePluginResponse<T> extends PluginResponse {
+  data: T;
+}
+
+export interface OnUpdateBatchPluginResponse<T> extends PluginResponse {
+  data: T[];
+}
+
+export interface BuildResultPluginResponse<T> extends PluginResponse {
+  data: T;
+}
+
+export interface RequestData {
+  state: any;
+  data: any;
+}
+
+export interface RequestDataBatch {
+  state: any;
+  data: any[];
+}
+
+export abstract class ComputedFieldPlugin<State, Data, Result> extends BasePlugin {
+  constructor() {
+    super();
+    this.initUpdateRoute();
+    this.initUpdateBatchRoute();
+    this.initBuildResultRoute();
+  }
+
+  abstract onUpdate(state: State | null, data: Data): State | null;
+
+  abstract buildResult(state: State | null): {
+    state: State | null;
+    result: Result;
+  };
+
+  private onUpdateBatch(state: State, data: Data[]): State | null {
```

Comment:
make it abstract and make the reduce logic before calling it in initUpdateBatchRoute